batan-server/engine/game.ts

```diff
@@ -749,15 +934,68 @@ export default class Game {
       let requiredNumberOfResources = resourceHarborOwned ? 2 : threeForOneHarborOwned ? 3 : 4;
       let playerResourceCount = this.players[playerIndex].getPlayerResource(resourceToBank) ?? 0;
       if (playerResourceCount >= requiredNumberOfResources) {
-        this.takeResources(player, resourceToBank, requiredNumberOfResources);
-        this.giveResources(player, resourceFromBank, 1);
+        if(this.giveResources(player, resourceFromBank, 1)){
+          this.takeResources(player, resourceToBank, requiredNumberOfResources);
+          return {
+            success: true,
+            reason: ""
+          }
+        } else {
+          return {
+            success: false,
+            reason: "Operation failed because bank is out of that resource."
+          }
+        }
       }
       else {
-        console.log("You don't have enough resources for this!");
+        return {
+          success: false,
+          reason: "Operation failed because the player didn't supply correct resources."
+        }
+      }
+    }
+  }
+
+  /**
+   * Adds a settlement in the setup phase.
+   * @param vertexId 
+   * @param owner 
+   */
+  addSettlementInSetup(vertexId: number, owner: player) {
+    if (this.board.addSettlementInSetup(vertexId, owner)) {
+      return {
+        success: true,
+        reason: ""
+      }
+    } else {
+      return {
+        success: true,
+        reason: "Operation failed due to illegal placement of settlement."
+      }
+    }
+  }
+
+  /**
+   * Adds a road in the setup phase.
+   * @param startVertexId 
+   * @param endVertexId 
+   * @param owner 
+   */
+  addRoadInSetup(startVertexId: number, endVertexId: number, owner: player) {
```

Comment:
Can we combine addRoadInSetup && addSettlementInSetup into an atomic operation? If one succeeds then the other fails, we could end up with some unfortunate board states